### PR TITLE
Fix the bug that affected the DAG end date.

### LINKED ISSUE #31555: Manually marking dag run state will change end date
### Apache Airflow version

main (development)

### What happened

May-25-2023 14-22-49

When manually marking a dag run as success or failure, the end date of the run will also change. This is most noticable in the grid view since the duration will immediately change.

This has been the case for a while.

### What you think should happen instead

We probably shouldn't alter the end_date.

### How to reproduce

Manually mark a dag run, especially an old one, and see the dag run duration bar charts change dramatically with one outlier, ruining the value of the rest of the chart.

### Operating System

any

### Versions of Apache Airflow Providers

_No response_

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by avkirilishin**
Closes: https://github.com/apache/airflow/issues/31555

The state management has been relocated from the `mark_task` module to the `DagRun`'s set state.
Changes to attributes were implemented in accordance with the following table:

| Old State \ New State| QUEUED                        | RUNNING                                             | SUCCESS                      | FAILED                       |
|----------------------|-------------------------------|-----------------------------------------------------|------------------------------|------------------------------|
| None                 | queued_at = timezone.utcnow() | if empty: start_date = timezone.utcnow()<br />end_date = None | end_date = timezone.utcnow() | end_date = timezone.utcnow() |
| QUEUED               | queued_at = timezone.utcnow() | if empty: start_date = timezone.utcnow()<br />end_date = None | end_date = timezone.utcnow() | end_date = timezone.utcnow() |
| RUNNING              | queued_at = timezone.utcnow()<br />start_date = None<br />end_date = None |                    | end_date = timezone.utcnow() | end_date = timezone.utcnow() |
| SUCCESS              | queued_at = timezone.utcnow()<br />start_date = None<br />end_date = None | start_date = timezone.utcnow()<br />end_date = None |                              |                              |
| FAILED               | queued_at = timezone.utcnow()<br />start_date = None<br />end_date = None | start_date = timezone.utcnow()<br />end_date = None |                              |                              |

Now it's look like:
screen-recording-2023-12-09-at-192842

### CODE REVIEW

**Comment by potiuk** [comment]
I think it would be great is some of us more familiar with the dates/scheduling/dag state would look at this one : It looks legit and I like the table description clarification, but I am not sure if I am the right one to validate it :)

File: airflow/models/dagrun.py [hunk 1]
```diff
@@ -272,11 +272,75 @@ def get_state(self):
         return self._state
 
     def set_state(self, state: DagRunState) -> None:
+        """Change the state of the DagRan.
```
Comment (ephraimbuddy) [inline airflow/models/dagrun.py:273]: `set_state` is only used in scheduler & backfill jobs. I don't see how your changes connect to manually marking dagrun success or failed.

Comment (avkirilishin) [reply airflow/models/dagrun.py:273]: > `set_state` is only used in scheduler & backfill jobs. I don't see how your changes connect to manually marking dagrun success or failed.

We use property with `set_state` [here](https://github.com/apache/airflow/blob/fd9d0b0f7de515b661ec459ba66778dc9730cf0f/airflow/models/dagrun.py#L347-L349). And we modify the state [here](https://github.com/apache/airflow/blob/fd9d0b0f7de515b661ec459ba66778dc9730cf0f/airflow/api/common/mark_tasks.py#L368).

Comment (potiuk) [reply airflow/models/dagrun.py:273]: Yep. It looks like it's a good idea to move date modification to `set_state` as it is the "write" of the state property. @ephraimbuddy - and maybe others who have a bit more insights and review, but it LGTM

**Review by potiuk — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

File: airflow/models/dagrun.py (hunk 1 above)
Comment (eladkal) [reply airflow/models/dagrun.py:273]: Talked with @ephraimbuddy he is OK with current code

**Pull request merged**